### ISSUE TITLE
Mechanic: Dynamic Energy Value Calculation

### ISSUE THREAD

**Issue by pahimar** [pahimar/Equivalent-Exchange#33]
The Dynamic Energy Value calculation system (DynEMC) will attempt to calculate the lossiest EMC value for an item, given various ways the item is created (recipes)

- The system will examine crafting recipes to build a dependency graph of items required to "craft" the item
  - These recipes could be traditional Shaped/Shapeless recipes, or more custom recipes (e.g. Thaumcraft crafting)
  - There will exist a may for modders to register their custom crafting recipes with Equivalent Exchange for the purpose of resolving EMC values that may otherwise indeterminable.
- Calculated EMC values should be deterministic (same value output, given same inputs)
- The system will require a certain number of seed values for the graph to resolve.
- Calculations should be done asynchronously so as not to tie up the main game thread.
- Calculations should be cached for later use (synchronized to disk), avoiding the need to recalculate them in different play sessions

**Comment by Drullkus** [comment]
Would there be way for mods to register handlers for their own special recipes?

**Comment by pahimar** [comment]
Yup! I'm envisioning it as being fairly similar to how it was done in earlier versions of the mod

Tracking ticket for the Recipe Registry: https://github.com/pahimar/Equivalent-Exchange/issues/52
Previous code: https://github.com/pahimar/Equivalent-Exchange-3/blob/master/src/main/java/com/pahimar/ee/api/recipe/RecipeRegistryProxy.java

**Comment by Drullkus** [comment]
Will this be also possible via IMC? To be able to set a soft EMC value for an item that's overridable if there is a lesser possible EMC value for the same item. Would you consider a hard override to force the EMC value for the item even though there are lower possible values? Also interested in custom NBT emc handlers.
